File: src/main/java/httpilot/HTTP.java
/*
 * Copyright 2011 Kazuhiro Sera
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
package httpilot;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;

public class HTTP {

	private HTTP() {
	}

	public static Response get(Request request) throws IOException {
		return request(Method.GET, request);
	}

	public static Response post(Request request) throws IOException {
		return request(Method.POST, request);
	}

	public static Response put(Request request) throws IOException {
		return request(Method.PUT, request);
	}

	public static Response delete(Request request) throws IOException {
		return request(Method.DELETE, request);
	}

	public static Response head(Request request) throws IOException {
		return request(Method.HEAD, request);
	}

	public static Response options(Request request) throws IOException {
		return request(Method.OPTIONS, request);
	}

	public static Response trace(Request request) throws IOException {
		return request(Method.TRACE, request);
	}

	public static Response request(Method method, Request request)
			throws IOException {

		HttpURLConnection conn = request.toHttpURLConnection(method);

		if (request.getCharset() != null) {
			conn.setRequestProperty("Accept-Charset", request.getCharset());
		}

		boolean needToThrowException = false;
		String exceptionMessage = "";
		InputStream is = null;
		try {
			// POST request message body
			if (request.getSpecifiedBody() != null) {
				conn.setDoOutput(true);
				conn.setRequestProperty("Content-Type",
						request.getSpecifiedContentType());
				OutputStream os = conn.getOutputStream();
				try {
					os.write(request.getSpecifiedBody());
				} finally {
					IO.close(os);
				}
			} else if (request.getFormParams() != null
					&& request.getFormParams().size() > 0) {
				conn.setDoOutput(true);
				conn.setRequestProperty("Content-Type",
						"application/x-www-form-urlencoded");
				byte[] body = request.getRequestBody()
						.asApplicationXWwwFormUrlencoded();
				OutputStream os = conn.getOutputStream();
				try {
					os.write(body);
				} finally {
					IO.close(os);
				}
			} else if (request.getMultipartFormData() != null
					&& request.getMultipartFormData().size() > 0) {
				String boundary = "----HTTPilotBoundary_"
						+ System.currentTimeMillis();
				conn.setRequestProperty("Content-Type",
						"multipart/form-data; boundary=" + boundary);
				byte[] body = request.getRequestBody().asMultipart(boundary);
				conn.setDoOutput(true);
				OutputStream os = conn.getOutputStream();
				try {
					os.write(body);
				} finally {
					IO.close(os);
				}
			}

			conn.connect();
			is = conn.getInputStream();

		} catch (IOException ioe) {
			if (request.isEnableThrowingIOException()) {
				needToThrowException = true;
				exceptionMessage = ioe.getMessage();
			}
			is = conn.getErrorStream();
		}

		Response response = new Response();
		response.setCharset(request.getCharset());
		response.setStatus(conn.getResponseCode());
		response.setHeaders(conn.getHeaderFields());
		if (is != null) {
			ByteArrayOutputStream os = new ByteArrayOutputStream();
			try {
				int c;
				while ((c = is.read()) != -1) {
					os.write(c);
				}
				response.setBody(os.toByteArray());
			} finally {
				IO.close(os);
			}
		}

		if (needToThrowException) {
			throw new HTTPIOException(exceptionMessage, response);
		} else {
			return response;
		}
	}
}

File: src/main/scala/httpilot/scala/HTTP.scala
/*
 * Copyright 2012 Kazuhiro Sera
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
package httpilot.scala

import scala.collection.JavaConverters._
import httpilot.{ HTTP => JavaHTTP, Request => JavaRequest }

object HTTP {

  def get(req: Request) = Response(JavaHTTP.get(req))

  def get(url: String, charset: String = httpilot.Request.DEFAULT_CHARSET) = {
    Response(JavaHTTP.get(new JavaRequest(url, charset)))
  }

  def post(req: Request) = Response(JavaHTTP.post(req))

  def post(url: String, data: String) = {
    Response(JavaHTTP.post(Request(url).body(data.getBytes)))
  }

  def post(url: String, formParams: Map[String, Any]) = {
    Response(JavaHTTP.post(new JavaRequest(url, formParams.map {
      case (k, v) => (k, v.asInstanceOf[java.lang.Object])
    }.asJava)))
  }

  def put(req: Request) = Response(JavaHTTP.put(req))

  def put(url: String, data: String) = {
    Response(JavaHTTP.put(new Request(url).body(data.getBytes)))
  }

  def put(url: String, formParams: Map[String, Any]) = {
    Response(JavaHTTP.put(new JavaRequest(url, formParams.map {
      case (k, v) => (k, v.asInstanceOf[java.lang.Object])
    }.asJava)))
  }

  def delete(req: Request) = Response(JavaHTTP.delete(req))

  def head(req: Request) = Response(JavaHTTP.head(req))

  def options(req: Request) = Response(JavaHTTP.options(req))

  def trace(req: Request) = Response(JavaHTTP.trace(req))

  def request(method: Method, req: Request) = Response(JavaHTTP.request(method, req))

}
